src/java.base/share/classes/java/nio/file/Path.java

```diff
@@ -248,6 +250,146 @@ public static Path of(URI uri) {
+            if (lastDotIndex > 0 && lastDotIndex < length - 1) {
+                return fileNameString.substring(lastDotIndex + 1);
+            }
+        }
+
+        return defaultExtension;
+    }
+
+    /**
+     * Returns whether the string representation of this path has an
+     * extension among the one or more extensions given as parameters.
+     * The extension of this path will be compared against the given
+     * parameters in the order in which they are supplied.
+     *
+     * @param   ext
+     *          the first extension for which to check
+     *
+     * @param   extensions
+     *          zero or more subsequent extensions for which to check
+     *
+     * @return  an {@code Optional} which either contains the found extension
+     *          or is empty if this path either does not have an extension or
+     *          its extension is not among those given
+     *
+     * @since 19
+     */
+    default Optional<String> hasExtension(String ext, String... extensions) {
+        String thisExtension = getExtension(null);
+        if (thisExtension == null)
+            return Optional.empty();
+
+        Objects.requireNonNull(ext);
+        if (ext.equals(thisExtension))
+            return Optional.of(ext);
+
+        for (String e : extensions) {
+            Objects.requireNonNull(e);
+            if (e.equals(thisExtension))
+                return Optional.of(e);
+        }
+
+        return Optional.empty();
+    }
+
+    /**
+     * Returns a new {@code Path} which is equal to this path aside from its
+     * extension which is replaced by the provided extension. If the provided
+     * extension is {@link String#isEmpty() empty}, then the current extension
+     * and the dot ('.') preceding it are removed. If this path does not have
+     * an extension then the provided extension is appended.
+     *
+     * @implNote
+     * This default implementation works with the string representation of this
+     * path. Paths which have a native representation should override this
+     * implementation in order to preserve that representation.
+     *
+     * @param  extension
+     *         the extension to replace this path's current extension;
+     *         if non-empty, may not contain a leading or trailing dot 
+     *         but is not otherwise checked for validity
+     *
+     * @return a new path equal to this path but with extension replaced by
+     *         that provided
+     *
+     * @throws IllegalArgumentException if the provided extension contains a
+     *         leading or trailing dot ('.')
+     *
+     * @since 19
+     */
+    default Path replaceExtension(String extension) {
+        Objects.requireNonNull(extension, "extension");
+
+        // verify the extension contains neither a leading nor trailing dot
+        if (!extension.isEmpty()) {
+            if (extension.indexOf('.') == 0 ||
+                extension.lastIndexOf('.') == extension.length() - 1)
```

Comment:
No need to scan the string using `indexOf`:

`(extension.charAt(0) == '.' || extension.charAt(extension.length()-1) == '.')`

Reply:
Commit 8ed8b1b25efa18a419d9176bff38f12616c9881d fixes this.